From 19e38d62cabaf37abcd515940304522fdede6f19 Mon Sep 17 00:00:00 2001
From: Gatien Chevallier <gatien.chevallier@foss.st.com>
Date: Tue, 28 May 2024 15:59:38 +0200
Subject: [PATCH 0913/1695] drivers: stm32_etzpc: use generic firewall API

Use the new firewall API to populate the firewall bus and register
the ETZPC as a firewall provider.

The ETZPC driver register the following ops:
-set_conf
-check_access
-acquire_access
-acquire_memory_access

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git f209d57003a59839f062ca6d55ec9571cae9559d drivers: stm32_etzpc: use generic firewall API]
Change-Id: I31ec68d42ddbbdb0b7e2648fc21b64055f5eac6e
Signed-off-by: Gatien Chevallier <gatien.chevallier@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/383374
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
Reviewed-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
Tested-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
---
 core/arch/arm/plat-stm32mp1/main.c  |  29 +-
 core/drivers/firewall/stm32_etzpc.c | 650 ++++++++++++----------------
 core/include/drivers/stm32_etzpc.h  |   9 -
 3 files changed, 293 insertions(+), 395 deletions(-)

diff --git a/core/arch/arm/plat-stm32mp1/main.c b/core/arch/arm/plat-stm32mp1/main.c
index 76ba52408..f541b48d1 100644
--- a/core/arch/arm/plat-stm32mp1/main.c
+++ b/core/arch/arm/plat-stm32mp1/main.c
@@ -9,11 +9,10 @@
 #include <console.h>
 #include <drivers/clk_dt.h>
 #include <drivers/counter.h>
+#include <drivers/firewall_device.h>
 #include <drivers/gic.h>
 #include <drivers/pinctrl.h>
 #include <drivers/stm32_bsec.h>
-#include <drivers/stm32_etzpc.h>
-#include <drivers/stm32_firewall.h>
 #include <drivers/stm32_gpio.h>
 #include <drivers/stm32_tamp.h>
 #include <drivers/stm32_uart.h>
@@ -371,21 +370,27 @@ service_init_late(init_stm32mp15_secure_srams);
 
 static TEE_Result init_stm32mp1_drivers(void)
 {
+	struct dt_driver_provider *prov = NULL;
 	TEE_Result res = TEE_ERROR_GENERIC;
-	const struct stm32_firewall_cfg sec_cfg[] = {
-		{ FWLL_SEC_RW | FWLL_MASTER(0) },
-		{ }, /* Null terminated */
+	uint32_t firewall_query_args[2] = { ETZPC_TZMA1_ID, SYSRAM_SEC_SIZE };
+	struct firewall_query firewall = {
+		.args = firewall_query_args,
+		.arg_count = ARRAY_SIZE(firewall_query_args),
 	};
+	int node = 0;
+
+	node = fdt_node_offset_by_compatible(get_embedded_dt(), -1,
+					     "st,stm32-etzpc");
+	if (node < 0)
+		panic("Could not get ETZPC node");
 
-	/* Without secure DTB support, some drivers must be inited */
-	if (!IS_ENABLED(CFG_EMBED_DTB))
-		stm32_etzpc_init(ETZPC_BASE);
+	prov = dt_driver_get_provider_by_node(node, DT_DRIVER_FIREWALL);
+	assert(prov);
 
-	res = stm32_firewall_set_config(ROM_BASE, ROM_SIZE, sec_cfg);
-	if (res)
-		panic("Unable to secure ROM");
+	firewall.ctrl = dt_driver_provider_priv_data(prov);
 
-	res = stm32_firewall_set_config(SYSRAM_BASE, SYSRAM_SEC_SIZE, sec_cfg);
+	/* Secure SYSRAM */
+	res = firewall_set_configuration(&firewall);
 	if (res)
 		panic("Unable to secure SYSRAM");
 
diff --git a/core/drivers/firewall/stm32_etzpc.c b/core/drivers/firewall/stm32_etzpc.c
index e918f78e4..cc8606df4 100644
--- a/core/drivers/firewall/stm32_etzpc.c
+++ b/core/drivers/firewall/stm32_etzpc.c
@@ -1,24 +1,19 @@
 // SPDX-License-Identifier: BSD-3-Clause
 /*
- * Copyright (c) 2015-2021, ARM Limited and Contributors. All rights reserved.
- * Copyright (c) 2017-2021, STMicroelectronics
+ * Copyright (c) 2015-2017, ARM Limited and Contributors. All rights reserved.
+ * Copyright (c) 2017-2024, STMicroelectronics
  */
 
 /*
  * STM32 ETPZC acts as a firewall on stm32mp SoC peripheral interfaces and
  * internal memories. The driver expects a single instance of the controller
  * in the platform.
- *
- * The driver API is defined in header file stm32_etzpc.h.
- *
- * Driver registers a PM callback for restoration of the access permissions
- * when it resumes.
  */
 
 #include <assert.h>
 #include <drivers/clk_dt.h>
 #include <drivers/stm32_etzpc.h>
-#include <drivers/stm32_firewall.h>
+#include <drivers/firewall.h>
 #include <drivers/stm32mp_dt_bindings.h>
 #include <initcall.h>
 #include <io.h>
@@ -28,8 +23,10 @@
 #include <kernel/panic.h>
 #include <kernel/pm.h>
 #include <kernel/spinlock.h>
+#include <kernel/tee_misc.h>
 #include <libfdt.h>
 #include <mm/core_memprot.h>
+#include <mm/core_mmu.h>
 #include <stm32_util.h>
 #include <util.h>
 
@@ -41,11 +38,11 @@
 #define ETZPC_ID_MASK			GENMASK_32(7, 0)
 
 /* ID Registers */
-#define ETZPC_TZMA0_SIZE		0x000U
-#define ETZPC_DECPROT0			0x010U
-#define ETZPC_DECPROT_LOCK0		0x030U
-#define ETZPC_HWCFGR			0x3F0U
-#define ETZPC_VERR			0x3F4U
+#define ETZPC_TZMA0_SIZE		U(0x000)
+#define ETZPC_DECPROT0			U(0x010)
+#define ETZPC_DECPROT_LOCK0		U(0x030)
+#define ETZPC_HWCFGR			U(0x3F0)
+#define ETZPC_VERR			U(0x3F4)
 
 /* ID Registers fields */
 #define ETZPC_TZMA0_SIZE_LOCK		BIT(31)
@@ -56,14 +53,12 @@
 #define ETZPC_HWCFGR_NUM_PER_SEC_SHIFT	8
 #define ETZPC_HWCFGR_NUM_AHB_SEC_MASK	GENMASK_32(23, 16)
 #define ETZPC_HWCFGR_NUM_AHB_SEC_SHIFT	16
-#define ETZPC_HWCFGR_CHUNCKS1N4_MASK	GENMASK_32(31, 24)
-#define ETZPC_HWCFGR_CHUNCKS1N4_SHIFT	24
+#define ETZPC_HWCFGR_CHUNKS1N4_MASK	GENMASK_32(31, 24)
+#define ETZPC_HWCFGR_CHUNKS1N4_SHIFT	24
 
 #define DECPROT_SHIFT			1
-#define IDS_PER_DECPROT_REGS		16U
-#define IDS_PER_DECPROT_LOCK_REGS	32U
-
-#define PAGE_SIZE			BIT(12) /* 4KB page size */
+#define IDS_PER_DECPROT_REGS		U(16)
+#define IDS_PER_DECPROT_LOCK_REGS	U(32)
 
 /*
  * Implementation uses uint8_t to store each securable DECPROT configuration
@@ -77,25 +72,27 @@
 
 /*
  * struct stm32_etzpc_platdata - Driver data set at initialization
- * @name - Name
- * @base - ETZPC IOMEM base address
- * @clk - ETZPC clock
- * @periph_cfg - Peripheral DECPROT configuraiton
- * @tzma_cfg - TZMA configuration
+ *
+ * @name:	Name of the peripheral
+ * @clk:	ETZPC clock
+ * @periph_cfg:	Peripheral DECPROT configuration
+ * @tzma_cfg:	TZMA configuration
+ * @base:	ETZPC IOMEM base address
  */
 struct stm32_etzpc_platdata {
 	const char *name;
-	uintptr_t base;
 	struct clk *clk;
 	uint8_t *periph_cfg;
 	uint16_t *tzma_cfg;
+	struct io_pa_va base;
 };
 
 /*
  * struct stm32_etzpc_driver_data - configuration data from the hardware
- * @num_tzma - number of TZMA zones, read from the hardware
- * @num_per_sec - number of securable AHB & APB periphs, read from the hardware
- * @num_ahb_sec - number of securable AHB master zones, read from the hardware
+ *
+ * @num_tzma:	 Number of TZMA zones, read from the hardware
+ * @num_per_sec: Number of securable AHB & APB periphs, read from the hardware
+ * @num_ahb_sec: Number of securable AHB master zones, read from the hardware
  */
 struct stm32_etzpc_driver_data {
 	unsigned int num_tzma;
@@ -104,19 +101,24 @@ struct stm32_etzpc_driver_data {
 };
 
 /*
- * struct etzpc_device - ETAPC device driver instance
- * @pdata - Platform data set during initialization
- * @ddata - Device configuration data from the hardware
- * @fdev - Firewall device instance registered for the ETZPC
- * @lock - Access contention
+ * struct etzpc_device - ETZPC device driver instance
+ * @pdata:	Platform data set during initialization
+ * @ddata:	Device configuration data from the hardware
+ * @lock:	Access contention
  */
 struct etzpc_device {
 	struct stm32_etzpc_platdata pdata;
-	struct stm32_etzpc_driver_data ddata;
-	struct stm32_firewall_device *fdev;
+	struct stm32_etzpc_driver_data *ddata;
 	unsigned int lock;
 };
 
+static const char *const etzpc_decprot_strings[] __maybe_unused = {
+	"ETZPC_DECPROT_S_RW",
+	"ETZPC_DECPROT_NS_R_S_W",
+	"ETZPC_DECPROT_MCU_ISOLATION",
+	"ETZPC_DECPROT_NS_RW",
+};
+
 static uint32_t etzpc_lock(struct etzpc_device *dev)
 {
 	return may_spin_lock(&dev->lock);
@@ -127,16 +129,15 @@ static void etzpc_unlock(struct etzpc_device *dev, uint32_t exceptions)
 	may_spin_unlock(&dev->lock, exceptions);
 }
 
-static bool __maybe_unused valid_decprot_id(struct etzpc_device *etzpc_dev,
-					    unsigned int id)
+static bool valid_decprot_id(struct etzpc_device *etzpc_dev, unsigned int id)
 {
-	return id < etzpc_dev->ddata.num_per_sec;
+	return id < etzpc_dev->ddata->num_per_sec;
 }
 
 static bool __maybe_unused valid_tzma_id(struct etzpc_device *etzpc_dev,
 					 unsigned int id)
 {
-	return id < etzpc_dev->ddata.num_tzma;
+	return id < etzpc_dev->ddata->num_tzma;
 }
 
 static enum etzpc_decprot_attributes etzpc_binding2decprot(uint32_t mode)
@@ -159,17 +160,18 @@ static enum etzpc_decprot_attributes etzpc_binding2decprot(uint32_t mode)
 
 static void etzpc_do_configure_decprot(struct etzpc_device *etzpc_dev,
 				       uint32_t decprot_id,
-				       enum etzpc_decprot_attributes decprot_attr)
+				       enum etzpc_decprot_attributes attr)
 {
-	size_t offset = 4U * (decprot_id / IDS_PER_DECPROT_REGS);
+	size_t offset = U(4) * (decprot_id / IDS_PER_DECPROT_REGS);
 	uint32_t shift = (decprot_id % IDS_PER_DECPROT_REGS) << DECPROT_SHIFT;
-	uint32_t masked_decprot = (uint32_t)decprot_attr & ETZPC_DECPROT0_MASK;
-	vaddr_t base = etzpc_dev->pdata.base;
+	uint32_t masked_decprot = (uint32_t)attr & ETZPC_DECPROT0_MASK;
+	vaddr_t base = etzpc_dev->pdata.base.va;
 	unsigned int exceptions = 0;
 
 	assert(valid_decprot_id(etzpc_dev, decprot_id));
 
-	FMSG("ID : %"PRIu32", CONF %d", decprot_id, decprot_attr);
+	FMSG("ID : %"PRIu32", CONF %s", decprot_id,
+	     etzpc_decprot_strings[attr]);
 
 	exceptions = etzpc_lock(etzpc_dev);
 
@@ -183,9 +185,9 @@ static void etzpc_do_configure_decprot(struct etzpc_device *etzpc_dev,
 static enum etzpc_decprot_attributes
 etzpc_do_get_decprot(struct etzpc_device *etzpc_dev, uint32_t decprot_id)
 {
-	size_t offset = 4U * (decprot_id / IDS_PER_DECPROT_REGS);
+	size_t offset = U(4) * (decprot_id / IDS_PER_DECPROT_REGS);
 	uint32_t shift = (decprot_id % IDS_PER_DECPROT_REGS) << DECPROT_SHIFT;
-	vaddr_t base = etzpc_dev->pdata.base;
+	vaddr_t base = etzpc_dev->pdata.base.va;
 	uint32_t value = 0;
 
 	assert(valid_decprot_id(etzpc_dev, decprot_id));
@@ -199,9 +201,9 @@ etzpc_do_get_decprot(struct etzpc_device *etzpc_dev, uint32_t decprot_id)
 static void etzpc_do_lock_decprot(struct etzpc_device *etzpc_dev,
 				  uint32_t decprot_id)
 {
-	size_t offset = 4U * (decprot_id / IDS_PER_DECPROT_LOCK_REGS);
+	size_t offset = U(4) * (decprot_id / IDS_PER_DECPROT_LOCK_REGS);
 	uint32_t mask = BIT(decprot_id % IDS_PER_DECPROT_LOCK_REGS);
-	vaddr_t base = etzpc_dev->pdata.base;
+	vaddr_t base = etzpc_dev->pdata.base.va;
 	uint32_t exceptions = 0;
 
 	assert(valid_decprot_id(etzpc_dev, decprot_id));
@@ -213,12 +215,12 @@ static void etzpc_do_lock_decprot(struct etzpc_device *etzpc_dev,
 	etzpc_unlock(etzpc_dev, exceptions);
 }
 
-static bool is_decprot_locked(struct etzpc_device *etzpc_dev,
+static bool decprot_is_locked(struct etzpc_device *etzpc_dev,
 			      uint32_t decprot_id)
 {
-	size_t offset = 4U * (decprot_id / IDS_PER_DECPROT_LOCK_REGS);
+	size_t offset = U(4) * (decprot_id / IDS_PER_DECPROT_LOCK_REGS);
 	uint32_t mask = BIT(decprot_id % IDS_PER_DECPROT_LOCK_REGS);
-	vaddr_t base = etzpc_dev->pdata.base;
+	vaddr_t base = etzpc_dev->pdata.base.va;
 
 	assert(valid_decprot_id(etzpc_dev, decprot_id));
 
@@ -229,7 +231,7 @@ static void etzpc_do_configure_tzma(struct etzpc_device *etzpc_dev,
 				    uint32_t tzma_id, uint16_t tzma_value)
 {
 	size_t offset = sizeof(uint32_t) * tzma_id;
-	vaddr_t base = etzpc_dev->pdata.base;
+	vaddr_t base = etzpc_dev->pdata.base.va;
 	uint32_t exceptions = 0;
 
 	assert(valid_tzma_id(etzpc_dev, tzma_id));
@@ -245,7 +247,7 @@ static uint16_t etzpc_do_get_tzma(struct etzpc_device *etzpc_dev,
 				  uint32_t tzma_id)
 {
 	size_t offset = sizeof(uint32_t) * tzma_id;
-	vaddr_t base = etzpc_dev->pdata.base;
+	vaddr_t base = etzpc_dev->pdata.base.va;
 
 	assert(valid_tzma_id(etzpc_dev, tzma_id));
 
@@ -255,7 +257,7 @@ static uint16_t etzpc_do_get_tzma(struct etzpc_device *etzpc_dev,
 static void etzpc_do_lock_tzma(struct etzpc_device *etzpc_dev, uint32_t tzma_id)
 {
 	size_t offset = sizeof(uint32_t) * tzma_id;
-	vaddr_t base = etzpc_dev->pdata.base;
+	vaddr_t base = etzpc_dev->pdata.base.va;
 	uint32_t exceptions = 0;
 
 	assert(valid_tzma_id(etzpc_dev, tzma_id));
@@ -267,10 +269,10 @@ static void etzpc_do_lock_tzma(struct etzpc_device *etzpc_dev, uint32_t tzma_id)
 	etzpc_unlock(etzpc_dev, exceptions);
 }
 
-static bool is_tzma_locked(struct etzpc_device *etzpc_dev, uint32_t tzma_id)
+static bool tzma_is_locked(struct etzpc_device *etzpc_dev, uint32_t tzma_id)
 {
 	size_t offset = sizeof(uint32_t) * tzma_id;
-	vaddr_t base = etzpc_dev->pdata.base;
+	vaddr_t base = etzpc_dev->pdata.base.va;
 
 	assert(valid_tzma_id(etzpc_dev, tzma_id));
 
@@ -281,10 +283,8 @@ static bool is_tzma_locked(struct etzpc_device *etzpc_dev, uint32_t tzma_id)
 static TEE_Result etzpc_pm(enum pm_op op, unsigned int pm_hint __unused,
 			   const struct pm_callback_handle *pm_handle)
 {
-	struct stm32_firewall_device *fdev =
-		(struct stm32_firewall_device *)pm_handle->handle;
-	struct etzpc_device *etzpc_dev = stm32_firewall_priv(fdev);
-	struct stm32_etzpc_driver_data *ddata = &etzpc_dev->ddata;
+	struct etzpc_device *etzpc_dev = pm_handle->handle;
+	struct stm32_etzpc_driver_data *ddata = etzpc_dev->ddata;
 	struct stm32_etzpc_platdata *pdata = &etzpc_dev->pdata;
 	unsigned int n = 0;
 
@@ -292,14 +292,14 @@ static TEE_Result etzpc_pm(enum pm_op op, unsigned int pm_hint __unused,
 		for (n = 0; n < ddata->num_per_sec; n++) {
 			pdata->periph_cfg[n] =
 				(uint8_t)etzpc_do_get_decprot(etzpc_dev, n);
-			if (is_decprot_locked(etzpc_dev, n))
+			if (decprot_is_locked(etzpc_dev, n))
 				pdata->periph_cfg[n] |= PERIPH_PM_LOCK_BIT;
 		}
 
 		for (n = 0; n < ddata->num_tzma; n++) {
 			pdata->tzma_cfg[n] =
 				(uint8_t)etzpc_do_get_tzma(etzpc_dev, n);
-			if (is_tzma_locked(etzpc_dev, n))
+			if (tzma_is_locked(etzpc_dev, n))
 				pdata->tzma_cfg[n] |= TZMA_PM_LOCK_BIT;
 		}
 
@@ -328,179 +328,226 @@ static TEE_Result etzpc_pm(enum pm_op op, unsigned int pm_hint __unused,
 
 	return TEE_SUCCESS;
 }
-
 DECLARE_KEEP_PAGER(etzpc_pm);
 
-static TEE_Result etzpc_mode_from_cfg(const struct stm32_firewall_cfg *cfg,
-				      enum etzpc_decprot_attributes *attr,
-				      bool *lock)
+static TEE_Result stm32_etzpc_check_access(struct firewall_query *firewall)
 {
-	const struct stm32_firewall_cfg *config = cfg;
-
-	assert(cfg && attr);
-
-	*attr = ETZPC_DECPROT_MAX;
-
-	while (config->access) {
-		uint32_t mode = 0;
-
-		mode = config->access & FWLL_CONF_MASK;
-		if (mode == FWLL_NO_ACCESS) {
-			config++;
-			continue;
-		}
-
-		if (lock && (config->access & FWLL_CONF_LOCK))
-			*lock |= true;
-
-		switch (config->access & FWLL_MASTER_MASK) {
-		case FWLL_MASTER(0):
-			if (*attr != ETZPC_DECPROT_MAX) {
-				DMSG("Inconsistent configuration data");
-				return TEE_ERROR_BAD_PARAMETERS;
-			}
-
-			if (mode & FWLL_SEC_RW)
-				*attr = ETZPC_DECPROT_S_RW;
+	struct etzpc_device *etzpc_dev = firewall->ctrl->priv;
+	enum etzpc_decprot_attributes attr_req = ETZPC_DECPROT_MAX;
+	enum etzpc_decprot_attributes attr = ETZPC_DECPROT_MAX;
+	uint32_t id = 0;
 
-			if (mode & FWLL_NSEC_READ)
-				*attr = ETZPC_DECPROT_NS_R_S_W;
+	if (!firewall || firewall->arg_count != 2)
+		return TEE_ERROR_BAD_PARAMETERS;
 
-			if (mode & FWLL_NSEC_WRITE)
-				*attr = ETZPC_DECPROT_NS_RW;
+	/*
+	 * Peripheral configuration, we assume the configuration is as
+	 * follows:
+	 * firewall->args[0]: Firewall ID
+	 * firewall->args[1]: DECPROT macro to extract etzpc_decprot_attributes
+	 *		      from
+	 */
+	id = firewall->args[0];
+	attr_req = etzpc_binding2decprot((firewall->args[1] >>
+					  ETZPC_MODE_SHIFT) & ETZPC_MODE_MASK);
+
+	if (id < etzpc_dev->ddata->num_per_sec) {
+		attr = etzpc_do_get_decprot(etzpc_dev, id);
+		DMSG("Check access %"PRIu32" - attr %s - requested %s", id,
+		     etzpc_decprot_strings[attr],
+		     etzpc_decprot_strings[attr_req]);
+
+		/*
+		 * Access authorized if the attributes requested match the
+		 * current configuration, or if the requester is secure and
+		 * the device is not MCU isolated, or if the requester is
+		 * non-secure and the device is not MCU isolated and not secure
+		 */
+		if (attr == attr_req ||
+		    ((attr_req == ETZPC_DECPROT_S_RW ||
+		      attr_req == ETZPC_DECPROT_NS_R_S_W) && attr !=
+		     ETZPC_DECPROT_MCU_ISOLATION) ||
+		    ((attr_req == ETZPC_DECPROT_NS_RW ||
+		      attr_req == ETZPC_DECPROT_NS_R_S_W) && attr !=
+		     ETZPC_DECPROT_MCU_ISOLATION && attr !=
+		     ETZPC_DECPROT_S_RW))
+			return TEE_SUCCESS;
+		else
+			return TEE_ERROR_ACCESS_DENIED;
+	} else {
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+}
 
-			break;
-#ifdef CFG_STM32MP15
-		case FWLL_MASTER(1):
-			if (*attr != ETZPC_DECPROT_MAX) {
-				DMSG("Inconsistent configuration data");
-				return TEE_ERROR_BAD_PARAMETERS;
-			}
+static TEE_Result stm32_etzpc_acquire_access(struct firewall_query *firewall)
+{
+	struct etzpc_device *etzpc_dev = firewall->ctrl->priv;
+	enum etzpc_decprot_attributes attr = ETZPC_DECPROT_MCU_ISOLATION;
+	uint32_t id = 0;
 
-			/* No other configuration set */
-			if (mode & FWLL_FULL_ACCESS &&
-			    *attr == ETZPC_DECPROT_MAX)
-				*attr = ETZPC_DECPROT_MCU_ISOLATION;
+	if (!firewall || firewall->arg_count != 2)
+		return TEE_ERROR_BAD_PARAMETERS;
 
-			break;
-#endif
-		default:
-			return TEE_ERROR_BAD_PARAMETERS;
-		}
+	id = firewall->args[0];
 
-		config++;
+	if (id < etzpc_dev->ddata->num_per_sec) {
+		attr = etzpc_do_get_decprot(etzpc_dev, id);
+		if (attr == ETZPC_DECPROT_MCU_ISOLATION)
+			return TEE_ERROR_ACCESS_DENIED;
+	} else {
+		return TEE_ERROR_BAD_PARAMETERS;
 	}
 
 	return TEE_SUCCESS;
 }
 
-static TEE_Result stm32_etzpc_has_access(struct stm32_firewall_device *fdev,
-					 unsigned int id,
-					 paddr_t base, size_t sz,
-					 const struct stm32_firewall_cfg *cfg)
+static TEE_Result
+stm32_etzpc_acquire_memory_access(struct firewall_query *firewall,
+				  paddr_t paddr, size_t size,
+				  bool read __unused, bool write __unused)
 {
-	struct etzpc_device *etzpc_dev = stm32_firewall_priv(fdev);
-	enum etzpc_decprot_attributes attr_req = ETZPC_DECPROT_MAX;
-	enum etzpc_decprot_attributes attr = ETZPC_DECPROT_MAX;
-	uint16_t size = 0;
-	uint32_t tzma_id = 0;
-	TEE_Result res = TEE_ERROR_GENERIC;
-
-	assert(cfg);
+	struct etzpc_device *etzpc_dev = firewall->ctrl->priv;
+	paddr_t tzma_base = 0;
+	size_t prot_size = 0;
+	uint32_t id = 0;
 
-	res = etzpc_mode_from_cfg(cfg, &attr_req, NULL);
-	if (res || attr_req == ETZPC_DECPROT_MAX)
-		return TEE_ERROR_SECURITY;
+	if (!firewall || firewall->arg_count != 2)
+		return TEE_ERROR_BAD_PARAMETERS;
 
-	if (id < etzpc_dev->ddata.num_per_sec) {
-		attr = etzpc_do_get_decprot(etzpc_dev, id);
-		FMSG("Check Access %u - attr %d - requested %d", id,
-		     attr, attr_req);
-		if (attr == attr_req)
-			return TEE_SUCCESS;
-		else
-			return TEE_ERROR_ACCESS_DENIED;
+	id = firewall->args[0];
+	switch (id) {
+	case ETZPC_TZMA0_ID:
+		tzma_base = ROM_BASE;
+		prot_size = etzpc_do_get_tzma(etzpc_dev, 0) * SMALL_PAGE_SIZE;
+		break;
+	case ETZPC_TZMA1_ID:
+		tzma_base = SYSRAM_BASE;
+		prot_size = etzpc_do_get_tzma(etzpc_dev, 1) * SMALL_PAGE_SIZE;
+		break;
+	default:
+		return TEE_ERROR_BAD_PARAMETERS;
 	}
 
-	/* TZMA */
-	tzma_id = (id - etzpc_dev->ddata.num_per_sec);
-	size = etzpc_do_get_tzma(etzpc_dev, tzma_id) * PAGE_SIZE;
-	if (base < fdev->compat->reg[id].addr + size &&
-	    base + sz - 1 < fdev->compat->reg[id].addr + size &&
-	    attr_req == ETZPC_DECPROT_S_RW)
-		return TEE_SUCCESS;
+	DMSG("Acquiring access for TZMA%u, secured from %#"PRIxPA" to %#"PRIxPA,
+	     id == ETZPC_TZMA0_ID ? 0 : 1, tzma_base, tzma_base + prot_size);
 
-	if (base >= fdev->compat->reg[id].addr + size &&
-	    base + sz - 1 < fdev->compat->reg[id].addr +
-	    fdev->compat->reg[id].size &&
-	    attr_req == ETZPC_DECPROT_NS_RW)
+	if (core_is_buffer_inside(paddr, size, tzma_base, prot_size))
 		return TEE_SUCCESS;
 
 	return TEE_ERROR_ACCESS_DENIED;
 }
 
-static TEE_Result stm32_etzpc_configure(struct stm32_firewall_device *fdev,
-					unsigned int id,
-					paddr_t base, size_t sz,
-					const struct stm32_firewall_cfg *cfg)
+static TEE_Result stm32_etzpc_configure(struct firewall_query *firewall)
 {
-	struct etzpc_device *etzpc_dev = stm32_firewall_priv(fdev);
+	struct etzpc_device *etzpc_dev = firewall->ctrl->priv;
 	enum etzpc_decprot_attributes attr = ETZPC_DECPROT_MAX;
-	TEE_Result res = TEE_ERROR_GENERIC;
 	unsigned int total_sz = 0;
-	uint32_t tzma_id = 0;
-	bool lock = false;
-
-	assert(cfg);
+	uint32_t id = 0;
 
-	res = etzpc_mode_from_cfg(cfg, &attr, &lock);
-	if (res || attr == ETZPC_DECPROT_MAX)
+	if (firewall->arg_count != 2)
 		return TEE_ERROR_BAD_PARAMETERS;
 
-	if (id < etzpc_dev->ddata.num_per_sec) {
-		if (is_decprot_locked(etzpc_dev, id))
+	id = firewall->args[0];
+
+	FMSG("Setting firewall configuration for peripheral ID: %"PRIu32, id);
+
+	if (id < etzpc_dev->ddata->num_per_sec) {
+		uint32_t mode = 0;
+
+		/*
+		 * Peripheral configuration, we assume the configuration is as
+		 * follows:
+		 * firewall->args[0]: Firewall ID
+		 * firewall->args[1]: Firewall configuration to apply
+		 */
+
+		mode = (firewall->args[1] >> ETZPC_MODE_SHIFT) &
+		       ETZPC_MODE_MASK;
+		attr = etzpc_binding2decprot(mode);
+
+		if (decprot_is_locked(etzpc_dev, id)) {
+			EMSG("Peripheral configuration locked");
 			return TEE_ERROR_ACCESS_DENIED;
+		}
 
 		etzpc_do_configure_decprot(etzpc_dev, id, attr);
-		if (lock)
+		if (firewall->args[1] & ETZPC_LOCK_MASK)
 			etzpc_do_lock_decprot(etzpc_dev, id);
 
 		return TEE_SUCCESS;
-	}
+	} else if (id == ETZPC_TZMA0_ID || id == ETZPC_TZMA1_ID) {
+		unsigned int tzma_id = 0;
+
+		/*
+		 * TZMA configuration, we assume the configuration is as
+		 * follows:
+		 * firewall->args[0]: One of ETZPC_TZMA0_ID/ETZPC_TZMA1_ID
+		 * firewall->args[1]: Memory size to secure
+		 */
+		switch (id) {
+		case ETZPC_TZMA0_ID:
+			if (firewall->args[1] > ROM_SIZE)
+				return TEE_ERROR_BAD_PARAMETERS;
 
-	/* TZMA */
-	if (base + sz > fdev->compat->reg[id].addr + fdev->compat->reg[id].size)
-		return TEE_ERROR_BAD_PARAMETERS;
+			tzma_id = 0;
+			break;
+		case ETZPC_TZMA1_ID:
+			if (firewall->args[1] > SYSRAM_SIZE)
+				return TEE_ERROR_BAD_PARAMETERS;
+
+			tzma_id = 1;
+			break;
+		default:
+			return TEE_ERROR_BAD_PARAMETERS;
+		}
+
+		assert(IS_ALIGNED(firewall->args[1], SMALL_PAGE_SIZE));
+
+		if (tzma_is_locked(etzpc_dev, tzma_id)) {
+			EMSG("TZMA configuration locked");
+			return TEE_ERROR_ACCESS_DENIED;
+		}
+
+		total_sz = ROUNDUP_DIV(firewall->args[1], SMALL_PAGE_SIZE);
+		etzpc_do_configure_tzma(etzpc_dev, tzma_id, total_sz);
+	} else {
+		EMSG("Unknown firewall ID: %"PRIu32, id);
 
-	switch (attr) {
-	case ETZPC_DECPROT_S_RW:
-		total_sz = base + sz - fdev->compat->reg[id].addr;
-		break;
-	case ETZPC_DECPROT_NS_RW:
-		total_sz = base - fdev->compat->reg[id].addr;
-		break;
-	default:
 		return TEE_ERROR_BAD_PARAMETERS;
 	}
 
-	total_sz = ROUNDUP_DIV(total_sz, PAGE_SIZE);
-	tzma_id = id - etzpc_dev->ddata.num_per_sec;
-	etzpc_do_configure_tzma(etzpc_dev, tzma_id, total_sz);
-	if (lock)
-		etzpc_do_lock_tzma(etzpc_dev, tzma_id);
-
 	return TEE_SUCCESS;
 }
 
-static const struct stm32_firewall_ops stm32_etzpc_fw_ops = {
-	.has_access = stm32_etzpc_has_access,
-	.configure_access = stm32_etzpc_configure,
-};
+static struct etzpc_device *stm32_etzpc_alloc(void)
+{
+	struct etzpc_device *etzpc_dev = calloc(1, sizeof(*etzpc_dev));
+	struct stm32_etzpc_driver_data *ddata = calloc(1, sizeof(*ddata));
+
+	if (etzpc_dev && ddata) {
+		etzpc_dev->ddata = ddata;
+		return etzpc_dev;
+	}
+
+	free(ddata);
+	free(etzpc_dev);
+
+	return NULL;
+}
+
+/* Informative unused function */
+static void stm32_etzpc_free(struct etzpc_device *etzpc_dev)
+{
+	if (etzpc_dev) {
+		free(etzpc_dev->ddata);
+		free(etzpc_dev);
+	}
+}
 
 static void stm32_etzpc_set_driverdata(struct etzpc_device *dev)
 {
-	struct stm32_etzpc_driver_data *ddata = &dev->ddata;
-	uintptr_t base = dev->pdata.base;
+	struct stm32_etzpc_driver_data *ddata = dev->ddata;
+	vaddr_t base = dev->pdata.base.va;
 	uint32_t reg = io_read32(base + ETZPC_HWCFGR);
 
 	ddata->num_tzma = (reg & ETZPC_HWCFGR_NUM_TZMA_MASK) >>
@@ -510,26 +557,20 @@ static void stm32_etzpc_set_driverdata(struct etzpc_device *dev)
 	ddata->num_ahb_sec = (reg & ETZPC_HWCFGR_NUM_AHB_SEC_MASK) >>
 			     ETZPC_HWCFGR_NUM_AHB_SEC_SHIFT;
 
-	DMSG("ETZPC revision 0x%02" PRIu8 ", per_sec %u, ahb_sec %u, tzma %u",
+	DMSG("ETZPC revision 0x%02"PRIx8", per_sec %u, ahb_sec %u, tzma %u",
 	     io_read8(base + ETZPC_VERR),
 	     ddata->num_per_sec, ddata->num_ahb_sec, ddata->num_tzma);
 }
 
-struct dt_id_attr {
-	/* The effective size of the array is meaningless here */
-	fdt32_t id_attr[1];
-};
-
 static void fdt_etzpc_conf_decprot(struct etzpc_device *dev,
 				   const void *fdt, int node)
 {
-	const struct dt_id_attr *conf_list = NULL;
+	const fdt32_t *cuint = NULL;
 	size_t i = 0;
 	int len = 0;
 
-	conf_list = (const struct dt_id_attr *)fdt_getprop(fdt, node,
-							   "st,decprot", &len);
-	if (!conf_list) {
+	cuint = fdt_getprop(fdt, node, "st,decprot", &len);
+	if (!cuint) {
 		DMSG("No ETZPC DECPROT configuration in DT");
 		return;
 	}
@@ -537,7 +578,7 @@ static void fdt_etzpc_conf_decprot(struct etzpc_device *dev,
 	clk_enable(dev->pdata.clk);
 
 	for (i = 0; i < len / sizeof(uint32_t); i++) {
-		uint32_t value = fdt32_to_cpu(conf_list->id_attr[i]);
+		uint32_t value = fdt32_to_cpu(cuint[i]);
 		uint32_t id = (value >> ETZPC_ID_SHIFT) & ETZPC_ID_MASK;
 		uint32_t mode = (value >> ETZPC_MODE_SHIFT) & ETZPC_MODE_MASK;
 		bool lock = value & ETZPC_LOCK_MASK;
@@ -568,14 +609,14 @@ static TEE_Result init_etzpc_from_dt(struct etzpc_device *etzpc_dev,
 
 	fdt_fill_device_info(fdt, &etzpc_info, node);
 	if (etzpc_info.reg == DT_INFO_INVALID_REG ||
-	    etzpc_info.reg_size == DT_INFO_INVALID_REG_SIZE ||
-	    etzpc_info.clock == DT_INFO_INVALID_CLOCK) {
+	    etzpc_info.reg_size == DT_INFO_INVALID_REG_SIZE)
 		return TEE_ERROR_ITEM_NOT_FOUND;
-	}
 
 	base.pa = etzpc_info.reg;
 	etzpc_dev->pdata.name = fdt_get_name(fdt, node, &len);
-	etzpc_dev->pdata.base = io_pa_or_va_secure(&base, etzpc_info.reg_size);
+	etzpc_dev->pdata.base.va = io_pa_or_va_secure(&base,
+						      etzpc_info.reg_size);
+	etzpc_dev->pdata.base = base;
 	res = clk_dt_get_by_index(fdt, node, 0, &etzpc_dev->pdata.clk);
 	if (res)
 		return res;
@@ -583,54 +624,69 @@ static TEE_Result init_etzpc_from_dt(struct etzpc_device *etzpc_dev,
 	stm32_etzpc_set_driverdata(etzpc_dev);
 
 	etzpc_dev->pdata.periph_cfg =
-		calloc(etzpc_dev->ddata.num_per_sec,
+		calloc(etzpc_dev->ddata->num_per_sec,
 		       sizeof(*etzpc_dev->pdata.periph_cfg));
+	if (!etzpc_dev->pdata.periph_cfg)
+		return TEE_ERROR_OUT_OF_MEMORY;
 
 	etzpc_dev->pdata.tzma_cfg =
-		calloc(etzpc_dev->ddata.num_tzma,
+		calloc(etzpc_dev->ddata->num_tzma,
 		       sizeof(*etzpc_dev->pdata.tzma_cfg));
-	if (!etzpc_dev->pdata.periph_cfg || !etzpc_dev->pdata.tzma_cfg)
+	if (!etzpc_dev->pdata.tzma_cfg) {
+		free(etzpc_dev->pdata.periph_cfg);
 		return TEE_ERROR_OUT_OF_MEMORY;
+	}
 
 	fdt_etzpc_conf_decprot(etzpc_dev, fdt, node);
 
 	return TEE_SUCCESS;
 }
 
+static const struct firewall_controller_ops firewall_ops = {
+	.set_conf = stm32_etzpc_configure,
+	.check_access = stm32_etzpc_check_access,
+	.acquire_access = stm32_etzpc_acquire_access,
+	.acquire_memory_access = stm32_etzpc_acquire_memory_access,
+};
+
 static TEE_Result stm32_etzpc_probe(const void *fdt, int node,
-				    const void *compat_data)
+				    const void *compat_data __unused)
 {
 	TEE_Result res = TEE_ERROR_GENERIC;
-	struct etzpc_device *etzpc_dev = NULL;
+	struct etzpc_device *etzpc_dev = stm32_etzpc_alloc();
+	struct firewall_controller *controller = NULL;
 
-	etzpc_dev = calloc(1, sizeof(*etzpc_dev));
 	if (!etzpc_dev) {
 		res = TEE_ERROR_OUT_OF_MEMORY;
 		goto err;
 	}
 
 	res = init_etzpc_from_dt(etzpc_dev, fdt, node);
-	if (res)
-		goto err;
+	if (res) {
+		stm32_etzpc_free(etzpc_dev);
+		return res;
+	}
 
-	etzpc_dev->fdev = stm32_firewall_dev_alloc();
-	if (!etzpc_dev->fdev) {
+	controller = calloc(1, sizeof(*controller));
+	if (!controller) {
 		res = TEE_ERROR_OUT_OF_MEMORY;
 		goto err;
 	}
 
-	etzpc_dev->fdev->name = etzpc_dev->pdata.name;
-	etzpc_dev->fdev->ops = &stm32_etzpc_fw_ops;
-	etzpc_dev->fdev->compat = (struct stm32_firewall_compat *)compat_data;
-	etzpc_dev->fdev->priv = etzpc_dev;
+	controller->base = &etzpc_dev->pdata.base;
+	controller->name = etzpc_dev->pdata.name;
+	controller->priv = etzpc_dev;
+	controller->ops = &firewall_ops;
 
-	res = stm32_firewall_dev_register(etzpc_dev->fdev);
+	res = firewall_dt_controller_register(fdt, node, controller);
 	if (res)
 		goto err;
 
-	stm32_firewall_bus_probe(etzpc_dev->fdev, fdt, node);
+	res = firewall_dt_probe_bus(fdt, node, controller);
+	if (res)
+		goto err;
 
-	register_pm_core_service_cb(etzpc_pm, etzpc_dev->fdev, "stm32-etzpc");
+	register_pm_core_service_cb(etzpc_pm, etzpc_dev, "stm32-etzpc");
 
 	return TEE_SUCCESS;
 
@@ -639,162 +695,8 @@ err:
 	panic();
 }
 
-#ifdef CFG_STM32MP13
-static struct stm32_firewall_reg
-etzpc_compat_reg_mp13[STM32MP1_ETZPC_MAX_ID + 2] = {
-	[STM32MP1_ETZPC_VREFBUF_ID] =		{ 0x50025000 },
-	[STM32MP1_ETZPC_LPTIM2_ID] =		{ 0x50021000 },
-	[STM32MP1_ETZPC_LPTIM3_ID] =		{ 0x50022000 },
-	[STM32MP1_ETZPC_LTDC_ID] =		{ 0x5A001000 },
-	[STM32MP1_ETZPC_DCMIPP_ID] =		{ 0x5A000000 },
-	[STM32MP1_ETZPC_USBPHYCTRL_ID] =	{ 0x5A006000 },
-	[STM32MP1_ETZPC_DDRCTRLPHY_ID] =	{ 0x5A003000 },
-	[STM32MP1_ETZPC_IWDG1_ID] =		{ 0x5C003000 },
-	[STM32MP1_ETZPC_STGENC_ID] =		{ 0x5C008000 },
-	[STM32MP1_ETZPC_USART1_ID] =		{ 0x4C000000 },
-	[STM32MP1_ETZPC_USART2_ID] =		{ 0x4C001000 },
-	[STM32MP1_ETZPC_SPI4_ID] =		{ 0x4C002000 },
-	[STM32MP1_ETZPC_SPI5_ID] =		{ 0x4C003000 },
-	[STM32MP1_ETZPC_I2C3_ID] =		{ 0x4C004000 },
-	[STM32MP1_ETZPC_I2C4_ID] =		{ 0x4C005000 },
-	[STM32MP1_ETZPC_I2C5_ID] =		{ 0x4C006000 },
-	[STM32MP1_ETZPC_TIM12_ID] =		{ 0x4C007000 },
-	[STM32MP1_ETZPC_TIM13_ID] =		{ 0x4C008000 },
-	[STM32MP1_ETZPC_TIM14_ID] =		{ 0x4C009000 },
-	[STM32MP1_ETZPC_TIM15_ID] =		{ 0x4C00A000 },
-	[STM32MP1_ETZPC_TIM16_ID] =		{ 0x4C00B000 },
-	[STM32MP1_ETZPC_TIM17_ID] =		{ 0x4C00C000 },
-	[STM32MP1_ETZPC_ADC1_ID] =		{ 0x48003000 },
-	[STM32MP1_ETZPC_ADC2_ID] =		{ 0x48004000 },
-	[STM32MP1_ETZPC_OTG_ID] =		{ 0x49000000 },
-	[STM32MP1_ETZPC_TSC_ID] =		{ 0x5000B000 },
-	[STM32MP1_ETZPC_RNG_ID] =		{ 0x54004000 },
-	[STM32MP1_ETZPC_HASH_ID] =		{ 0x54003000 },
-	[STM32MP1_ETZPC_CRYP_ID] =		{ 0x54002000 },
-	[STM32MP1_ETZPC_SAES_ID] =		{ 0x54005000 },
-	[STM32MP1_ETZPC_PKA_ID] =		{ 0x54006000 },
-	[STM32MP1_ETZPC_BKPSRAM_ID] =		{ 0x54000000, 0x4000 },
-	[STM32MP1_ETZPC_ETH1_ID] =		{ 0x5800A000 },
-	[STM32MP1_ETZPC_ETH2_ID] =		{ 0x5800E000 },
-	[STM32MP1_ETZPC_SDMMC1_ID] =		{ 0x58005000 },
-	[STM32MP1_ETZPC_SDMMC2_ID] =		{ 0x58007000 },
-	[STM32MP1_ETZPC_MCE_ID] =		{ 0x58001000 },
-	[STM32MP1_ETZPC_FMC_ID] =		{ 0x58002000 },
-	[STM32MP1_ETZPC_QSPI_ID] =		{ 0x58003000 },
-	[STM32MP1_ETZPC_SRAM1_ID] =		{ 0x30000000, 0x4000 },
-	[STM32MP1_ETZPC_SRAM2_ID] =		{ 0x30004000, 0x2000 },
-	[STM32MP1_ETZPC_SRAM3_ID] =		{ 0x30006000, 0x2000 },
-	/* MAX ID to catch the TZMA address */
-	[STM32MP1_ETZPC_MAX_ID] =		{ 0x00000000, 0x20000},
-	[STM32MP1_ETZPC_MAX_ID + 1] =		{ 0x2FFE0000, 0x20000},
-};
-
-static const struct stm32_firewall_compat etzpc_compat[] = {
-	{
-		.reg = etzpc_compat_reg_mp13,
-		.compat_size = ARRAY_SIZE(etzpc_compat_reg_mp13)
-	},
-};
-#endif
-
-#ifdef CFG_STM32MP15
-static struct stm32_firewall_reg
-etzpc_compat_reg_mp15[STM32MP1_ETZPC_MAX_ID + 2] = {
-	[STM32MP1_ETZPC_STGENC_ID] =		{ 0x5C008000 },
-	[STM32MP1_ETZPC_BKPSRAM_ID] =		{ 0x54000000, 0x1000 },
-	[STM32MP1_ETZPC_IWDG1_ID] =		{ 0x5C003000 },
-	[STM32MP1_ETZPC_USART1_ID] =		{ 0x5C000000 },
-	[STM32MP1_ETZPC_SPI6_ID] =		{ 0x5C001000 },
-	[STM32MP1_ETZPC_I2C4_ID] =		{ 0x5C002000 },
-	[STM32MP1_ETZPC_RNG1_ID] =		{ 0x54003000 },
-	[STM32MP1_ETZPC_HASH1_ID] =		{ 0x54002000 },
-	[STM32MP1_ETZPC_CRYP1_ID] =		{ 0x54001000 },
-	[STM32MP1_ETZPC_DDRCTRL_ID] =		{ 0x5A003000 },
-	[STM32MP1_ETZPC_DDRPHYC_ID] =		{ 0x5A004000 },
-	[STM32MP1_ETZPC_I2C6_ID] =		{ 0x5C009000 },
-	[STM32MP1_ETZPC_TIM2_ID] =		{ 0x40000000 },
-	[STM32MP1_ETZPC_TIM3_ID] =		{ 0x40001000 },
-	[STM32MP1_ETZPC_TIM4_ID] =		{ 0x40002000 },
-	[STM32MP1_ETZPC_TIM5_ID] =		{ 0x40003000 },
-	[STM32MP1_ETZPC_TIM6_ID] =		{ 0x40004000 },
-	[STM32MP1_ETZPC_TIM7_ID] =		{ 0x40005000 },
-	[STM32MP1_ETZPC_TIM12_ID] =		{ 0x40006000 },
-	[STM32MP1_ETZPC_TIM13_ID] =		{ 0x40007000 },
-	[STM32MP1_ETZPC_TIM14_ID] =		{ 0x40008000 },
-	[STM32MP1_ETZPC_LPTIM1_ID] =		{ 0x40009000 },
-	[STM32MP1_ETZPC_WWDG1_ID] =		{ 0x4000A000 },
-	[STM32MP1_ETZPC_SPI2_ID] =		{ 0x4000B000 },
-	[STM32MP1_ETZPC_SPI3_ID] =		{ 0x4000C000 },
-	[STM32MP1_ETZPC_SPDIFRX_ID] =		{ 0x4000D000 },
-	[STM32MP1_ETZPC_USART2_ID] =		{ 0x4000E000 },
-	[STM32MP1_ETZPC_USART3_ID] =		{ 0x4000F000 },
-	[STM32MP1_ETZPC_UART4_ID] =		{ 0x40010000 },
-	[STM32MP1_ETZPC_UART5_ID] =		{ 0x40011000 },
-	[STM32MP1_ETZPC_I2C1_ID] =		{ 0x40012000 },
-	[STM32MP1_ETZPC_I2C2_ID] =		{ 0x40013000 },
-	[STM32MP1_ETZPC_I2C3_ID] =		{ 0x40014000 },
-	[STM32MP1_ETZPC_I2C5_ID] =		{ 0x40015000 },
-	[STM32MP1_ETZPC_CEC_ID] =		{ 0x40016000 },
-	[STM32MP1_ETZPC_DAC_ID] =		{ 0x40017000 },
-	[STM32MP1_ETZPC_UART7_ID] =		{ 0x40018000 },
-	[STM32MP1_ETZPC_UART8_ID] =		{ 0x40019000 },
-	[STM32MP1_ETZPC_MDIOS_ID] =		{ 0x4001C000 },
-	[STM32MP1_ETZPC_TIM1_ID] =		{ 0x44000000 },
-	[STM32MP1_ETZPC_TIM8_ID] =		{ 0x44001000 },
-	[STM32MP1_ETZPC_USART6_ID] =		{ 0x44003000 },
-	[STM32MP1_ETZPC_SPI1_ID] =		{ 0x44004000 },
-	[STM32MP1_ETZPC_SPI4_ID] =		{ 0x44005000 },
-	[STM32MP1_ETZPC_TIM15_ID] =		{ 0x44006000 },
-	[STM32MP1_ETZPC_TIM16_ID] =		{ 0x44007000 },
-	[STM32MP1_ETZPC_TIM17_ID] =		{ 0x44008000 },
-	[STM32MP1_ETZPC_SPI5_ID] =		{ 0x44009000 },
-	[STM32MP1_ETZPC_SAI1_ID] =		{ 0x4400A000 },
-	[STM32MP1_ETZPC_SAI2_ID] =		{ 0x4400B000 },
-	[STM32MP1_ETZPC_SAI3_ID] =		{ 0x4400C000 },
-	[STM32MP1_ETZPC_DFSDM_ID] =		{ 0x4400D000 },
-	[STM32MP1_ETZPC_TT_FDCAN_ID] =		{ 0x4400E000 },
-	[STM32MP1_ETZPC_LPTIM2_ID] =		{ 0x50021000 },
-	[STM32MP1_ETZPC_LPTIM3_ID] =		{ 0x50022000 },
-	[STM32MP1_ETZPC_LPTIM4_ID] =		{ 0x50023000 },
-	[STM32MP1_ETZPC_LPTIM5_ID] =		{ 0x50024000 },
-	[STM32MP1_ETZPC_SAI4_ID] =		{ 0x50027000 },
-	[STM32MP1_ETZPC_VREFBUF_ID] =		{ 0x50025000 },
-	[STM32MP1_ETZPC_DCMI_ID] =		{ 0x4C006000 },
-	[STM32MP1_ETZPC_CRC2_ID] =		{ 0x4C004000 },
-	[STM32MP1_ETZPC_ADC_ID] =		{ 0x48003000 },
-	[STM32MP1_ETZPC_HASH2_ID] =		{ 0x4C002000 },
-	[STM32MP1_ETZPC_RNG2_ID] =		{ 0x4C003000 },
-	[STM32MP1_ETZPC_CRYP2_ID] =		{ 0x4C005000 },
-	[STM32MP1_ETZPC_SRAM1_ID] =		{ 0x10000000, 0x20000 },
-	[STM32MP1_ETZPC_SRAM2_ID] =		{ 0x10020000, 0x20000 },
-	[STM32MP1_ETZPC_SRAM3_ID] =		{ 0x10040000, 0x10000 },
-	[STM32MP1_ETZPC_SRAM4_ID] =		{ 0x10050000, 0x10000 },
-	[STM32MP1_ETZPC_RETRAM_ID] =		{ 0x38000000, 0x10000 },
-	[STM32MP1_ETZPC_OTG_ID] =		{ 0x49000000 },
-	[STM32MP1_ETZPC_SDMMC3_ID] =		{ 0x48004000 },
-	[STM32MP1_ETZPC_DLYBSD3_ID] =		{ 0x48005000 },
-	[STM32MP1_ETZPC_DMA1_ID] =		{ 0x48000000 },
-	[STM32MP1_ETZPC_DMA2_ID] =		{ 0x48001000 },
-	[STM32MP1_ETZPC_DMAMUX_ID] =		{ 0x48002000 },
-	[STM32MP1_ETZPC_FMC_ID] =		{ 0x58002000 },
-	[STM32MP1_ETZPC_QSPI_ID] =		{ 0x58003000 },
-	[STM32MP1_ETZPC_DLYBQ_ID] =		{ 0x58004000 },
-	[STM32MP1_ETZPC_ETH_ID] =		{ 0x5800A000 },
-	/* MAX ID to catch the TZMA address */
-	[STM32MP1_ETZPC_MAX_ID] =		{ 0x00000000, 0x20000},
-	[STM32MP1_ETZPC_MAX_ID + 1] =		{ 0x2FFC0000, 0x40000},
-};
-
-static const struct stm32_firewall_compat etzpc_compat[] = {
-	{
-		.reg = etzpc_compat_reg_mp15,
-		.compat_size = ARRAY_SIZE(etzpc_compat_reg_mp15)
-	},
-};
-#endif
-
 static const struct dt_device_match etzpc_match_table[] = {
-	{ .compatible = "st,stm32-etzpc", .compat_data = etzpc_compat },
+	{ .compatible = "st,stm32-etzpc", },
 	{ }
 };
 
diff --git a/core/include/drivers/stm32_etzpc.h b/core/include/drivers/stm32_etzpc.h
index d5ae46eff..15493a5a9 100644
--- a/core/include/drivers/stm32_etzpc.h
+++ b/core/include/drivers/stm32_etzpc.h
@@ -17,13 +17,4 @@ enum etzpc_decprot_attributes {
 	ETZPC_DECPROT_NS_RW = 3,
 	ETZPC_DECPROT_MAX = 4,
 };
-
-#define ETZPC_TZMA_ALL_SECURE		GENMASK_32(9, 0)
-#define ETZPC_TZMA_ALL_NO_SECURE	0x0
-
-#ifdef CFG_STM32_ETZPC
-void stm32_etzpc_init(paddr_t base);
-#else
-static inline void stm32_etzpc_init(paddr_t __unused base) {}
-#endif
 #endif /*__DRIVERS_STM32_ETZPC_H*/
-- 
2.47.2

